Addition of stochastic spread model and random geometric model

MWolfR opened the pull request:
## Description

Added two stochastic graph models:
 - A random geometric model with some custom modifications.
 - The stochastic spread model. Combined with the above this yields the SGSG model

Added to modeling a function that fits per-node biases to a reference ConnectivityMatrix.

MWolfR committed:
Added stochastic spread model

MWolfR committed:
Added random geometric graph with customizations

MWolfR committed:
Improved formatting and added per group bias fitting.

danielaegassan commented on src/connalysis/modelling/modelling.py:
```diff
@@ -1412,3 +1412,47 @@ def _plot_3rd_order(adj, node_properties, model_name, p_conn_dist_bip, count_con
+
+    Parameters
+    ---------
+    reference : conntility.ConnectivityMatrix
```
This doesn't follow our convention.  We wanted to make the package agnostic to different ways to wrap matrices. So the inputs are always an adjacency matrix and a data frame (possibly empty). In particular, connectome-utilities is not a dependency of connalysis

Please follow the convention so that it fits with the rest of the code.

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -126,3 +127,185 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
+
+def _evaluate_probs_less_random(p_mat, adjust=None):
+    """
+    Evaluates a single spread step using the `less random` method.
```
This description of the method is not clear, please clarify.

danielaegassan replied:
@MWolfR this doc is missing the output

danielaegassan commented on src/connalysis/randomization/randomization.py:
```diff
@@ -753,3 +762,309 @@ def add_connections(adj,nc, seed=0,sparse_mode=True, max_iter=30):
+                            custom_w_out=None, custom_w_in=None,
+                            no_diag=True, mirror=False):
+    """
+    Generates a directed random geometric graph with some optional modifications. The modifications introduce
```
We need to make more precise what we mean here, since there are many things called a random geometric graph. Maybe something like:

Generates a directed random geometric graph with optional modifications.

A basic random geometric graph is constructed from nodes embedded in R^m, with positions specified by `pts`. Directed edges between nodes are created probabilistically based on their pairwise distances in the ambient space.  The connection probability can be defined in multiple ways, depending on the input parameters provided. 

Additional modifications can be introduced to generate biases with respect to which nodes will be connected.

danielaegassan replied:
New docstring suggested.

MWolfR committed:
Addressing PR requests.

See: https://github.com/openbraininstitute/connectome-analysis/pull/14

MWolfR committed:
Hotfix: mirror must be False if pts_x is provided!

MWolfR commented:
@danielaegassan I think I added all the requested changes. Can you re-review, please?

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -126,3 +127,190 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
+        (see above) this is 1 / r.
+    less_random : bool
+        If set to true, then the less stochastic version of the spread is used. That is, for each source node 
+        (row in p_mat) the process spreads to exactly the expected number of nodes instead of a random number.
```
@MWolfR this docstring is missing the output

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -126,3 +127,190 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
+        If set to True, the output matrix is made symmetrical with the following strategy: If a connection
+        from vertex i to j exists, a connection is also placed from j to i if it does not already exist.
+    shape : tuple
+        The shape of the output matrix. If not provided, it will be guessed based on the indices in `w`.
```
@MWolfR this docstring is missing the output

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -126,3 +127,190 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
+        Directionality bias cannot be combined with distance bias (see below).
+    distance_func : function
+        A function that is to be evaluated on pairwise distances of candidate pairs. The function takes a
+        distance as input and returns a relative weight. 
```
@MWolfR this docstring is missing the output

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -126,3 +127,190 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
+        Shape: (n, ). Specifies the locations of potential edges. The array has one entry per source vertex,
+        i.e., the same length as `w_out`.
+        Each entry is a list of indices into `w_in` that specifies the potential targets of that source vertex.
+        This corresponds to the output obtained from querying a scipy.spatial.KDTree.
```
@MWolfR this docstring is missing the output

danielaegassan commented on src/connalysis/randomization/randomization.py:
```diff
@@ -753,3 +762,316 @@ def add_connections(adj,nc, seed=0,sparse_mode=True, max_iter=30):
+    directionality_fac : float
+        Must be between -1 and 1. This specifies how much the dot product calculated using `directionality_axis`
+        (see above) is weighed to calculate a bias. Formally:
+        w = (delta o directionality_axis) * directionality_fac + 1,
```
@MWolfR what is 

delta o directionality_axis 

?

MWolfR replied:
Literally the next line of the docstring: 
`where o denotes the vector dot product, delta is the vector from the source to the target vertex of a potential connection.`

danielaegassan replied:
sorry, I missed that ... all good

danielaegassan commented on src/connalysis/randomization/randomization.py:
```diff
@@ -753,3 +762,316 @@ def add_connections(adj,nc, seed=0,sparse_mode=True, max_iter=30):
+        Note that these are relative weights that are scaled to fulfill the constraints
+        on the number of connections to pick given by `n_pick` or `p_pick`.
+        Note that this does NOT affect the selection of candidates, parameterized by `n_neighbors` or 
+        `dist_neighbors`, but only the selection from the candidates.
```
Lines 828-829 are confusing.  What does this mean, if I choose 10 NN and then I bias I only end up with a subset of this given by the bias?

MWolfR replied:
Kind of. 
The _size_ of the subset is determined by `n_pick` or `p_pick`. Usually: picked with uniform probabilities. This bias only changes that from uniform to non-uniform probabilities (here: based on the orientation of the pair).

I do not know how to make this clearer, please suggest a way.

danielaegassan commented on src/connalysis/randomization/randomization.py:
```diff
@@ -753,3 +762,316 @@ def add_connections(adj,nc, seed=0,sparse_mode=True, max_iter=30):
+    sparse.csc_matrix
+        The adjacency matrix of an instance of a random geometric graph model.
+    
+    Raises
```
@MWolfR how does this work if I chose the nearest neighbours with p_pick? I don't get the logic.

MWolfR replied:
Think of it as a succession of filters:
  - You begin with all pairs
  - They are filtered based on pairwise distances. This is parameterized by `n_neighbors` _xor_ `dist_neighbors`.
  - The results are filtered once more based on `n_pick` _xor_ `p_pick`. This filtering can be further biased (i.e., moving to non-uniform probabilities) using `directionality_fac` _and_ `directionality_axis`,  _xor_ `distance_func`.

danielaegassan replied:
I resolved this, and basically added this conversation to the doc string.

danielaegassan commented on src/connalysis/randomization/randomization.py:
```diff
@@ -753,3 +762,316 @@ def add_connections(adj,nc, seed=0,sparse_mode=True, max_iter=30):
+        determined number. This avoids a large number of source nodes with zero out-degree. Set to 0 to
+        not use this feature.
+    q : float 
+        Sets the expected number of nodes to spread to in each step. This is done by scaling the weights in
```
@MWolfR what does the dynamic spread mean in the case of M of type bool?

MWolfR replied:
Clarified in new version

danielaegassan reviewed (commented):
@MWolfR a few more comments on the lines I didn't manage to check before.  Please review.

MWolfR committed:
Renamed functions to ensure they show up in docs. Clarified some docstrings. Added new "node_can_spread" optional argument.

danielaegassan commented on src/connalysis/randomization/rand_utils.py:
```diff
@@ -129,7 +129,7 @@ def add_bidirectional_connections(sparse_matrix: sp.csc_matrix, bedges_to_add: i
-def _evaluate_probs_less_random(p_mat, adjust=None):
+def evaluate_probs_less_random(p_mat, adjust=None):
     """
     Evaluates a single spread step using the `less random` method.
```
@MWolfR in this all all the other functions with similar comment, the point was not to remove the _ from before the function, but rather that the docstrings don't have a return part where it says what it's returned. For some of them is kind of obvious but for others it's not. So just for good practice is good to have them in all.

danielaegassan committed:
Change description of RGG 

I think what would make things clearer is a more detailed overall explanation of the function. Check if it makes sense to you.

MWolfR committed:
Added return statements to docstrings.
